src/App/Fossa/Main.hs

```diff
@@ -34,29 +36,47 @@ import Data.Text (Text)
 import qualified Data.Text as T
 import Discovery.Filters (BuildTargetFilter (..), filterParser)
 import Effect.Logger
+import Effect.ReadFS
 import Fossa.API.Types (ApiKey(..), ApiOpts(..))
 import Options.Applicative
 import System.Environment (lookupEnv)
 import System.Exit (die)
 import qualified System.Info as SysInfo
 import Text.Megaparsec (errorBundlePretty, runParser)
-import Text.URI (URI)
-import Text.URI.QQ (uri)
+import Text.URI (URI, mkURI)
 import Path
 
 windowsOsName :: String
 windowsOsName = "mingw32"
 
 mainPrefs :: ParserPrefs
-mainPrefs = prefs $ mconcat 
+mainPrefs = prefs $ mconcat
   [ helpShowGlobals,
     showHelpOnError,
-    subparserInline 
+    subparserInline
   ]
 
+mergeFileCmdConfig :: CmdOptions -> ConfigFile -> CmdOptions
+mergeFileCmdConfig cmd file =
+  CmdOptions
+    { optDebug = optDebug cmd
+    , optBaseUrl = optBaseUrl cmd <|> (configServer file >>= mkURI)
+    , optProjectName = optProjectName cmd <|> (configProject file >>= configProjID)
+    , optProjectRevision = optProjectRevision cmd <|> (configRevision file >>= configCommit)
+    , optAPIKey = optAPIKey cmd <|> configApiKey file
+    , optCommand = optCommand cmd
+    }
+
 appMain :: IO ()
 appMain = do
-  CmdOptions {..} <- customExecParser mainPrefs (info (opts <**> helper) (fullDesc <> header "fossa-cli - Flexible, performant dependency analysis"))
+  cmdConfig <- customExecParser mainPrefs (info (opts <**> helper) (fullDesc <> header "fossa-cli - Flexible, performant dependency analysis"))
+  configRead <- Diag.runDiagnosticsIO $ runReadFSIO readConfigFile
```

Comment:
One reason I did this is that I needed to be able to `die` in the main function since it returns `IO ()` as opposed to the `Maybe Config` that `readConfigFile` returns. Do you have another suggestion for working around this? I'm not entirely sure how to only handle the effects within the `Configuration` file unless I can `die` inside of it.

Reply:
Repeating from our paring - `die` returns `IO a`, since there's no reason it should be restricted to unit types only.  That should help redefine this logic in a separate function.  It's also worth pointing out that the type of `die` here is already monomorphized to `String -> IO (Maybe ConfigFile)`, otherwise ghc could not resolve it between the `Left` and `Right` branches of the case statement.